crates/sui-core/src/authority/transaction_reject_vote_reason_cache.rs

```diff
@@ -0,0 +1,185 @@
+///   fast path, as it relies on consensus position (epoch, block, index) to uniquely identify transactions
+/// - **Pre-consensus Rejections**: Direct rejections during transaction submission (before consensus
+///   propagation) are not cached since these transactions never enter the consensus pipeline
+/// - **Automatic Cleanup**: Maintains a retention period based on the last committed leader round
+///   and automatically purges older entries to prevent unbounded memory growth
+///
+/// ## Use Cases:
+/// - Providing detailed rejection reasons to clients querying transaction status
+/// - Debugging transaction failures in the fast path voting process
+pub(crate) struct TransactionRejectVoteReasonCache {
+    cache: RwLock<BTreeMap<ConsensusPosition, SuiError>>,
+    retention_rounds: u32,
+    epoch: EpochId,
+}
+
+impl TransactionRejectVoteReasonCache {
+    pub fn new(retention_rounds: Option<u32>, epoch: EpochId) -> Self {
+        Self {
+            cache: Default::default(),
+            retention_rounds: retention_rounds.unwrap_or(DEFAULT_RETENTION_ROUNDS),
+            epoch,
+        }
+    }
+
+    /// Records a rejection vote reason for a transaction at the specified consensus position. The consensus `position` that
+    /// uniquely identifies the transaction and the `reason` (SuiError) that caused the transaction to be rejected during voting
+    /// should be provided.
+    pub fn set_rejection_vote_reason(&self, position: ConsensusPosition, reason: &SuiError) {
+        debug_assert_eq!(position.epoch, self.epoch, "Epoch mismatch");
+        self.cache.write().insert(position, reason.clone());
+    }
+
+    /// Returns the rejection vote reason for the transaction at the specified consensus position. The result will be `None` when:
+    /// * this node has never casted a reject vote for the transaction in question (either accepted or not processed it).
+    /// * the transaction vote reason has been cleaned up due to the retention policy.
+    pub fn get_rejection_vote_reason(&self, position: ConsensusPosition) -> Option<SuiError> {
+        debug_assert_eq!(position.epoch, self.epoch, "Epoch mismatch");
+        self.cache.read().get(&position).cloned()
+    }
+
+    /// Sets the last committed leader round. This is used to clean up the cache based on the retention policy.
+    pub fn set_last_committed_leader_round(&self, round: u32) {
+        let _scope =
+            monitored_scope("TransactionRejectVoteReasonCache::set_last_committed_leader_round");
+        let cut_off_round = round.saturating_sub(self.retention_rounds) + 1;
+        let cut_off_position = ConsensusPosition {
+            epoch: self.epoch,
+            block: BlockRef::new(cut_off_round, AuthorityIndex::MIN, BlockDigest::MIN),
+            index: TransactionIndex::MIN,
+        };
+
+        let mut cache = self.cache.write();
+        let remaining = cache.split_off(&cut_off_position);
```

Comment:
I believe the clean up will be quite cheap so it can happen synchronously , but let me know if you feel otherwise